tests/waku_relay/test_protocol.nim

```diff
@@ -0,0 +1,1303 @@
+      
+      # When sending the 1KiB message
+      handlerFuture = newPushHandlerFuture()
+      otherHandlerFuture = newPushHandlerFuture()
+      discard await node.publish(pubsubTopic, msg1)
+
+      # Then the message is received in both nodes
+      check:
+        await handlerFuture.withTimeout(3.seconds)
+        await otherHandlerFuture.withTimeout(3.seconds)
+        (pubsubTopic, msg1) == handlerFuture.read()
+        (pubsubTopic, msg1) == otherHandlerFuture.read()
+      
+      # When sending the 10KiB message
+      handlerFuture = newPushHandlerFuture()
+      otherHandlerFuture = newPushHandlerFuture()
+      discard await node.publish(pubsubTopic, msg2)
+
+      # Then the message is received in both nodes
+      check:
+        await handlerFuture.withTimeout(3.seconds)
+        await otherHandlerFuture.withTimeout(3.seconds)
+        (pubsubTopic, msg2) == handlerFuture.read()
+        (pubsubTopic, msg2) == otherHandlerFuture.read()
+      
+      # When sending the 100KiB message
+      handlerFuture = newPushHandlerFuture()
+      otherHandlerFuture = newPushHandlerFuture()
+      discard await node.publish(pubsubTopic, msg3)
+
+      # Then the message is received in both nodes
+      check:
+        await handlerFuture.withTimeout(3.seconds)
+        await otherHandlerFuture.withTimeout(3.seconds)
+        (pubsubTopic, msg3) == handlerFuture.read()
+        (pubsubTopic, msg3) == otherHandlerFuture.read()
+      
+      # When sending the 1023KiB message
+      handlerFuture = newPushHandlerFuture()
+      otherHandlerFuture = newPushHandlerFuture()
+      discard await node.publish(pubsubTopic, msg4)
+
+      # Then the message is received in both nodes
+      check:
+        await handlerFuture.withTimeout(3.seconds)
+        await otherHandlerFuture.withTimeout(3.seconds)
+        (pubsubTopic, msg4) == handlerFuture.read()
+        (pubsubTopic, msg4) == otherHandlerFuture.read()
+
+      # When sending the 1024KiB message
+      handlerFuture = newPushHandlerFuture()
+      otherHandlerFuture = newPushHandlerFuture()
+      discard await node.publish(pubsubTopic, msg5)
+
+      # Then the message is received in both nodes
+      check:
+        await handlerFuture.withTimeout(3.seconds)
+        not await otherHandlerFuture.withTimeout(3.seconds)
+        (pubsubTopic, msg5) == handlerFuture.read()
+
+      # When sending the 1025KiB message
+      handlerFuture = newPushHandlerFuture()
+      otherHandlerFuture = newPushHandlerFuture()
+      discard await node.publish(pubsubTopic, msg6)
+
+      # Then the message is received in both nodes
+      check:
+        await handlerFuture.withTimeout(3.seconds)
+        not await otherHandlerFuture.withTimeout(3.seconds)
+        (pubsubTopic, msg6) == handlerFuture.read()
+
+      # Finally stop the other node
+      await allFutures(otherSwitch.stop(), otherNode.stop())
+
+    asyncTest "Multiple messages at once":
```

Comment:
Are you actually testing the protocol or testing your handler (and whether it successfully adds to `seq`)?

Because this is not "sending messages at once", this is sending them sequentially and the only change from other tests is custom handled adding to seq

Reply:
Unfortunately, this is the closest I managed to get to safely testing "send at once" because of an issue with the future (used for assertions) in the handler. The problem is calling `complete` on a `completed` future causes a `SEGFAULT`.

I found two ways of proceeding with this test:

1. Remove the future. This is the ideal solution. With this I'd be able to "send at once" (without any awaits) all messages, and after that, check whether they exist in the `seq`. The problem is, because of asynchronicity, I need to await an arbitrary number of seconds (which is not ideal for tests). Also, I can't remember off the top of my head, but for some reason, this didn't work sometimes.

2. Add the future. This works flawlessly, but it doesn't actually check for "send at once" because, as you mentioned, we await before sending each messages.

Any suggestion?

Reply:
I can't, because there's just future, and it needs to be awaited and resetted sequentially. Otherwise, you'd be calling `complete` on a `completed` future, which triggers the aforementioned `SEGFAULT`.